tests/trx_generator/trx_provider.hpp

```diff
@@ -42,63 +42,64 @@ namespace eosio::testing {
    };
 
-   struct null_trx_generator {
-      void generate_and_send() {}
-   };
 
-   constexpr int64_t min_sleep_us = 100;
+   constexpr int64_t min_sleep_us = 1;
 
    template<typename G, typename M>
    struct trx_tps_tester {
-      G _generator;
-      M _monitor;
+      std::shared_ptr<G> _generator;
+      std::shared_ptr<M>_monitor;
 
-      uint32_t _gen_duration_seconds;
-      uint32_t _target_tps;
+      int64_t _gen_duration_seconds;
+      int64_t _target_tps;
 
-      trx_tps_tester(G generator, M monitor, uint32_t gen_duration_seconds, uint32_t target_tps) :
-            _generator(), _monitor(), _gen_duration_seconds(gen_duration_seconds), _target_tps(target_tps) {
+      trx_tps_tester(std::shared_ptr<G> generator, std::shared_ptr<M> monitor, uint32_t gen_duration_seconds, uint32_t target_tps) :
+            _generator(generator), _monitor(monitor),
+               _gen_duration_seconds(gen_duration_seconds), _target_tps(target_tps) {
 
       }
 
       void run() {
          tps_test_stats stats;
+         fc::microseconds trx_interval(std::chrono::microseconds(1s).count() / _target_tps);
+         ilog("transaction interval = ${trxi}", ("trxi", trx_interval.count()));
 
          stats.total_trxs = _gen_duration_seconds * _target_tps;
          stats.trxs_left = stats.total_trxs;
          stats.start_time = fc::time_point::now();
          stats.expected_end_time = stats.start_time + fc::microseconds{_gen_duration_seconds * std::chrono::microseconds(1s).count()};
+         stats.last_sleep = 0;
 
          bool keep_running = true;
-         fc::microseconds trx_interval{std::chrono::microseconds(1s).count() / _target_tps};
-
-         fc::time_point last_run;
-         fc::time_point next_run;
 
          while (keep_running) {
-            last_run = fc::time_point::now();
-            next_run = last_run + trx_interval;
+            stats.last_run = fc::time_point::now();
+            stats.next_run = stats.start_time + fc::microseconds(trx_interval.count() * (stats.trxs_sent+1));
 
-            _generator.generate_and_send();
+            _generator->generate_and_send();
             stats.trxs_left--;
+            stats.trxs_sent++;
 
-            keep_running = (_monitor.monitor_test(stats) && stats.trxs_left);
+            keep_running = (_monitor->monitor_test(stats) && stats.trxs_left);
 
             if (keep_running) {
-               fc::microseconds time_to_sleep{next_run - fc::time_point::now()};
-               if (time_to_sleep.count() > min_sleep_us) {
+               fc::microseconds time_to_sleep{stats.next_run - fc::time_point::now()};
+               if (time_to_sleep.count() >= min_sleep_us) {
+                  stats.last_sleep = time_to_sleep.count();
                   std::this_thread::sleep_for(std::chrono::microseconds(time_to_sleep.count()));
+               } else {
+                  stats.last_sleep = time_to_sleep.count();
```

Comment:
This could be moved out of `if` and remove `else`.

Reply:
Actually that is a copy pasta error. It should be: 
stats.last_sleep = 0

Reply:
Actually, I take that back.  I am using that value in the test to determine that the tps tester knows that it is behind schedule.  I updated the member name to properly reflect the meaning of the value.